### PR TITLE
Move back to using listeners for Navigation events

### PR DESCRIPTION
**Pull request by narenkmanoharan**
Also fixed tests that were broken by previous PRs

### CODE REVIEW

File: magellan-legacy/src/main/java/com/wealthfront/magellan/Navigator.kt [hunk 1]
```diff
@@ -40,17 +40,6 @@ public class Navigator internal constructor(
     }
   }
 
-  public fun addLifecycleListener(screenLifecycleListener: ScreenLifecycleListener) {
```
Comment (narenkmanoharan) [inline magellan-legacy/src/main/java/com/wealthfront/magellan/Navigator.kt:43]: I don't think we need this anymore and we can just attach it directly to the `NavigationPropogator`

Comment (narenkmanoharan) [inline magellan-legacy/src/main/java/com/wealthfront/magellan/Navigator.kt:43]: I don't think we need this anymore and we can just attach it directly to the `NavigationPropogator`

File: magellan-legacy/src/test/java/com/wealthfront/magellan/LegacyViewComponentTest.java [hunk 2]
```diff
@@ -37,7 +37,6 @@ public void viewComponentLifecycleUntilDestroy() {
     legacyViewComponent.create(context);
 
     assertThat(screen.getView()).isEqualTo(null);
-    assertThat(screen.getActivity()).isEqualTo(context);
```
Comment (narenkmanoharan) [inline magellan-legacy/src/test/java/com/wealthfront/magellan/LegacyViewComponentTest.java:40]: sneaky. No more activity access in onCreate

File: magellan-library/src/main/java/com/wealthfront/magellan/navigation/NavigationDelegate.kt [hunk 3]
```diff
@@ -57,7 +53,6 @@ public class NavigationDelegate(
   }
 
   override fun onDestroy(context: Context) {
-    backStack.navigables().forEach { removeFromLifecycle(it) }
```
Comment (narenkmanoharan) [inline magellan-library/src/main/java/com/wealthfront/magellan/navigation/NavigationDelegate.kt:60]: we already remove these navigables before this step so this actually crashes when we do it again.

File: magellan-legacy/src/main/java/com/wealthfront/magellan/Navigator.kt (hunk 1 above)
Comment (ryanmoelter) [reply magellan-legacy/src/main/java/com/wealthfront/magellan/Navigator.kt:43]: Should we kept it for backwards compatibility?

File: magellan-library/src/main/java/com/wealthfront/magellan/navigation/CurrentNavigableProvider.kt [hunk 4]
```diff
@@ -2,35 +2,28 @@ package com.wealthfront.magellan.navigation
...
-import kotlinx.coroutines.launch
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 @OpenForMocking
-public class CurrentNavigableProvider @Inject constructor(
-  createdScope: CreatedLifecycleScope
-) : LifecycleAwareComponent() {
+public class CurrentNavigableProvider @Inject constructor() : NavigationListener, LifecycleAwareComponent() {
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/navigation/CurrentNavigableProvider.kt:11]: nit: probably can make this just `LifecycleAware` now

File: magellan-library/src/main/java/com/wealthfront/magellan/navigation/LoggingNavigableListener.kt [hunk 5]
```diff
@@ -2,50 +2,37 @@ package com.wealthfront.magellan.navigation
...
 
 @Singleton
 public class LoggingNavigableListener @Inject constructor(
-  private val navigationTraverser: NavigationTraverser,
-  createdScope: CreatedLifecycleScope
-) : LifecycleAwareComponent() {
-
-  private val createdScope by lifecycle(createdScope)
+  private val navigationTraverser: NavigationTraverser
+) : NavigationListener, LifecycleAwareComponent() {
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/navigation/LoggingNavigableListener.kt:13]: Same here, can probably just make this LifecycleAware

File: magellan-library/src/main/java/com/wealthfront/magellan/navigation/NavigationPropagator.kt [hunk 6]
```diff
@@ -1,23 +1,40 @@
...
+  public fun addNavigableListener(navigationListener: NavigationListener) {
+    listeners = listeners + navigationListener
+  }
+
+  @JvmStatic
+  public fun removeNavigableListener(navigationListener: NavigationListener) {
+    listeners = listeners - navigationListener
+  }
+
+  internal fun beforeNavigation() {
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/navigation/NavigationPropagator.kt:17]: Were these internal before? Idk but I like it :)

File: magellan-library/src/test/java/com/wealthfront/magellan/navigation/NavigationTraverserTest.kt [hunk 7]
```diff
@@ -76,26 +76,6 @@ class NavigationTraverserTest {
     )
   }
 
-  @Test
-  fun globalBackStackWithSiblingJourney() {
```
Comment (ryanmoelter) [inline magellan-library/src/test/java/com/wealthfront/magellan/navigation/NavigationTraverserTest.kt:80]: Why did we get rid of this test?

**Review by ryanmoelter — APPROVED** [review]
LGTM overall!

File: magellan-library/src/main/java/com/wealthfront/magellan/navigation/LoggingNavigableListener.kt (hunk 5 above)
Comment (narenkmanoharan) [reply magellan-library/src/main/java/com/wealthfront/magellan/navigation/LoggingNavigableListener.kt:13]: changed

File: magellan-library/src/main/java/com/wealthfront/magellan/navigation/CurrentNavigableProvider.kt (hunk 4 above)
Comment (narenkmanoharan) [reply magellan-library/src/main/java/com/wealthfront/magellan/navigation/CurrentNavigableProvider.kt:11]: done

File: magellan-library/src/test/java/com/wealthfront/magellan/navigation/NavigationTraverserTest.kt (hunk 7 above)
Comment (narenkmanoharan) [reply magellan-library/src/test/java/com/wealthfront/magellan/navigation/NavigationTraverserTest.kt:80]: Not sure. will add back

File: magellan-legacy/src/main/java/com/wealthfront/magellan/Navigator.kt (hunk 1 above)
Comment (narenkmanoharan) [reply magellan-legacy/src/main/java/com/wealthfront/magellan/Navigator.kt:43]: Actually, yeah. I can add it back.

**Pull request merged**